### PR TITLE
React Native Weather Sensor Library

### PR DESCRIPTION
**Pull request by ClaireLundy**
- added native libraries as git modules 
- used React Native Modules to bridge between native libraries and React
- tested adding to another React Native project using yalc

### CODE REVIEW

File: lefthook.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# EXAMPLE USAGE
```
Comment (paigebrinks) [inline lefthook.yml]: is this file necessary?

File: .gitmodules [hunk 2]
```diff
@@ -0,0 +1,8 @@
+[submodule "WeatherSensorLibAndroid"]
+	path = android/spindance.code-challenge.mobile.android.weather-sensor-service
+	url = https://github.com/spindance/spindance.code-challenge.mobile.android.weather-sensor-service.git
+	branch = main
+[submodule "WeatherSensorLibIos"]
+	path = ios/spindance.code-challenge.mobile.ios.weather-sensor-service-package
+	url = https://github.com/spindance/spindance.code-challenge.mobile.ios.weather-sensor-service-package.git
+	branch = main
```
Comment (ambers-spindance) [inline .gitmodules]: Did you actually get the submodules working then? By which I mean, the native code is written in their respective modules, and not in this repo?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
+# Contributing
+
+We want this community to be friendly and respectful to each other. Please follow it in all your interactions with the project.
```
Comment (ambers-spindance) [inline CONTRIBUTING.md]: I'd say maybe rephrase this a little since "it" in this context isn't really referencing anything which makes the statement unclear.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+## Development workflow
+
+To get started with the project, run `yarn` in the root directory to install the required dependencies for each package:
```
Comment (ambers-spindance) [inline CONTRIBUTING.md]: just `yarn`, not `yarn install`? (genuine question)

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+## Code of Conduct
+
+### Our Pledge
+
+We as members, contributors, and leaders pledge to make participation in our community a harassment-free experience for everyone, regardless of age, body size, visible or invisible disability, ethnicity, sex characteristics, gender identity and expression, level of experience, education, socio-economic status, nationality, personal appearance, race, religion, or sexual identity and orientation.
+
+We pledge to act and interact in ways that contribute to an open, welcoming, diverse, inclusive, and healthy community.
+
+### Our Standards
```
Comment (ambers-spindance) [inline CONTRIBUTING.md]: Is this all because the library is/will be public?

File: android/spindance.code-challenge.mobile.android.weather-sensor-service/LICENSE.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2022 SpinDance
```
Comment (ambers-spindance) [inline android/spindance.code-challenge.mobile.android.weather-sensor-service/LICENSE.md]: License appears twice? Is that related to the submodules thing?

**Review by ambers-spindance — COMMENTED** [review]
Mostly just some questions. Definitely have someone else look at it too, some of this stuff I just don't know enough about to tell what's happening.

File: .gitmodules (hunk 2 above)
Comment (ClaireLundy) [reply .gitmodules]: Yep, I had to figure out how to ignore/include specific files from the repos to get it to work

File: CONTRIBUTING.md (hunk 3 above)
Comment (ClaireLundy) [reply CONTRIBUTING.md]: Ah, I totally meant to delete this file. It just came with the auto-generated library but I don't think we need it

File: CONTRIBUTING.md (hunk 4 above)
Comment (ClaireLundy) [reply CONTRIBUTING.md]: Yes, the library is set up so that you can run just yarn and it will run yarn install and pod install at the same time.

File: android/spindance.code-challenge.mobile.android.weather-sensor-service/LICENSE.md (hunk 6 above)
Comment (ClaireLundy) [reply android/spindance.code-challenge.mobile.android.weather-sensor-service/LICENSE.md]: Hum maybe, I'll fix that

**Review by paigebrinks — APPROVED** [review]
Looks good to me. Just one file I think you can remove

File: README.md [hunk 7]
```diff
@@ -1 +1,13 @@
-# spindance.code-challenge.mobile.react-native.weather-sensor-service
\ No newline at end of file
+# react-native-weather-sensor-module
+react-native wrapper of native weather sensor generation libraries
```
Comment (roberthartman) [inline README.md]: Should you capitalize "react-native"?

File: README.md [hunk 8]
```diff
@@ -1 +1,13 @@
...
+## Installation
```
Comment (roberthartman) [inline README.md:3]: Should there be a blank line above line 3, or does it look fine when rendered?

File: README.md [hunk 9]
```diff
@@ -1 +1,13 @@
...
+You will need to publish the library locally using yalc
+
+1. Globally install yalc `npm install -g yalc`
+2. Clone this repo and `npm install`
+3. Publish the library `yalc publish`
+4. Add to your your project: 
+  1. `yalc add react-native-weather-sensor-module`
+  2. `npm install`
+
```
Comment (roberthartman) [inline README.md]: Extra blank line

**Comment by roberthartman** [comment]
@kevin-r-wilson can you take a quick look at this to make sure the Android side of things looks good to you (gradle versions etc)?

File: android/spindance.code-challenge.mobile.android.weather-sensor-service/.gitattributes [hunk 10]
```diff
@@ -0,0 +1,6 @@
+#
+# https://help.github.com/articles/dealing-with-line-endings/
+#
+# These are explicitly windows files and should use crlf
+*.bat           text eol=crlf
+
```
Comment (roberthartman) [inline android/spindance.code-challenge.mobile.android.weather-sensor-service/.gitattributes]: Shouldn't the `android` directory (where this file is) be pretty much empty, since it's a submodule? This includes the license, .gitignore, readme, gradle files and source code (*.kt) below.

Comment (roberthartman) [reply android/spindance.code-challenge.mobile.android.weather-sensor-service/.gitattributes]: OK, maybe I didn't look close enough, sorry. It looks like this si the Kotlin and Swift code you had to write in order to bridge things appropriately. My bad.

Comment (roberthartman) [reply android/spindance.code-challenge.mobile.android.weather-sensor-service/.gitattributes]: ... nevertheless, should the readme and license files be present here?

File: ios/WeatherSensorModule.swift [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import Foundation
+import Combine
+
+@available(iOS 13.0, *)
+@objc(WeatherSensorModule)
+class WeatherSensorModule: RCTEventEmitter {
+    private var cancellable : Any?
+
+    public static var shared:WeatherSensorModule?
```
Comment (roberthartman) [inline ios/WeatherSensorModule.swift]: There needs to be a space after `:`

File: android/src/main/java/com/reactnativeweathersensormodule/WeatherSensorModulePackage.kt [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+class WeatherSensorModulePackage : ReactPackage {
+    override fun createNativeModules(reactContext: ReactApplicationContext): List<NativeModule> {
+        return listOf(WeatherSensorModule(reactContext))
+    }
+
+    override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> {
+        return emptyList()
+    }
+}
```
Comment (roberthartman) [inline android/src/main/java/com/reactnativeweathersensormodule/WeatherSensorModulePackage.kt:16]: This file looks like it is using 4 space indentation, whereas the previous kt file is using 2. They should be consistent.

File: ios/WeatherSensorModule.swift [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+        override init() {
+            super.init()
+            WeatherSensorModule.shared = self
+        }
+        
+        override func supportedEvents() -> [String]! {
+            return ["reading"]
+        }
```
Comment (roberthartman) [inline ios/WeatherSensorModule.swift]: Extra indentation lines 11-18

Comment (roberthartman) [reply ios/WeatherSensorModule.swift]: Inconsistent indentation below as well (2, below 4 spaces above)

File: ios/WeatherSensorModule.swift [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    
+  @available(iOS 13.0, *)
+  @objc
+  func collectReadings() -> Void {  
+    cancellable = WeatherSensorService.shared.reader.sensorReadingsPublisher
+      .sink{value in
```
Comment (roberthartman) [inline ios/WeatherSensorModule.swift]: There needs to be a space after the `{`

File: react-native-weather-sensor-module.podspec [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+        "CLANG_CXX_LANGUAGE_STANDARD" => "c++17"
+    }
+
+    s.dependency "React-Codegen"
+    s.dependency "RCT-Folly", folly_version
+    s.dependency "RCTRequired"
+    s.dependency "RCTTypeSafety"
+    s.dependency "ReactCommon/turbomodule/core"
+  end
+end
```
Comment (roberthartman) [inline react-native-weather-sensor-module.podspec:35]: Just curious - why is there a podspec? How does Cocoapods fit into this? I'd think that would only come into play for iOS native.

File: tsconfig.build.json [hunk 16]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (roberthartman) [inline tsconfig.build.json]: Leading blank line seems icky, maybe it's just me :-)

**Review by roberthartman — CHANGES_REQUESTED** [review]
A little clean up needed in the native code, and I put down some questions for things I wasn't sure about.

File: android/spindance.code-challenge.mobile.android.weather-sensor-service/.gitattributes (hunk 10 above)
Comment (ClaireLundy) [reply android/spindance.code-challenge.mobile.android.weather-sensor-service/.gitattributes]: When I added the git submodules, it just cloned the entire repo into the android folder. Since these files were included in the android weather sensor repo they are included here.

File: react-native-weather-sensor-module.podspec (hunk 15 above)
Comment (ClaireLundy) [reply react-native-weather-sensor-module.podspec:35]: I am under the assumption that it is need to point to the ios repo. So when I add this library to another project it reads the podspec and knows where to look for additional ios code. It is needed for other repos to find the ios code.

File: android/spindance.code-challenge.mobile.android.weather-sensor-service/.gitattributes (hunk 10 above)
Comment (roberthartman) [reply android/spindance.code-challenge.mobile.android.weather-sensor-service/.gitattributes]: On your local hard drive, the submodule will look like a clone, but in git, it's really just a reference to the submodule. So I wouldn't expect there to be any files from the submodule itself committed in this repo.

File: react-native-weather-sensor-module.podspec (hunk 15 above)
Comment (roberthartman) [reply react-native-weather-sensor-module.podspec:35]: So maybe the RN Xcode project points at the Swift package via a cocoapod? Man that seems weird to me.

Comment (ClaireLundy) [reply react-native-weather-sensor-module.podspec:35]: Ya that sounds right. This was part of the auto generated code from [react-native-builder-bob](https://github.com/callstack/react-native-builder-bob). Not sure why they did it that way.

File: android/spindance.code-challenge.mobile.android.weather-sensor-service/.gitattributes (hunk 10 above)
Comment (ClaireLundy) [reply android/spindance.code-challenge.mobile.android.weather-sensor-service/.gitattributes]: Oh I see what you are saying, I'll look into this

Comment (ClaireLundy) [reply android/spindance.code-challenge.mobile.android.weather-sensor-service/.gitattributes]: I re-added the modules and now the folders just link to the respective repos which seems must more like the expected behavior

File: ios/WeatherSensorModule.swift [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+  public static var shared: WeatherSensorModule?
+    override init() {
+      super.init()
+      WeatherSensorModule.shared = self
+    }
+        
+    override func supportedEvents() -> [String]! {
+      return ["reading"]
+    }
```
Comment (roberthartman) [inline ios/WeatherSensorModule.swift]: There is mixed indentation level on lines 9-17 (2 vs 4 spaces).

File: ios/WeatherSensorModule.swift [hunk 18]
```diff
@@ -0,0 +1,46 @@
+import Foundation
+import Combine
+
+@available(iOS 13.0, *)
+@objc(WeatherSensorModule)
+class WeatherSensorModule: RCTEventEmitter {
+  private var cancellable : Any?
+
+  public static var shared: WeatherSensorModule?
+    override init() {
```
Comment (roberthartman) [inline ios/WeatherSensorModule.swift]: Should be a blank line before this function (after line 9)

Comment (roberthartman) [reply ios/WeatherSensorModule.swift]: Is the caller meant to be able to create an instance of this class, or that are intended to use the `shared` singleton? I assume the latter, in which case the init should be private. Now maybe you can't do that due to something I am not thinking of. But if can make the init private, you should.

File: ios/WeatherSensorModule.swift [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+        
+    override func supportedEvents() -> [String]! {
+      return ["reading"]
+    }
+    
+  @available(iOS 13.0, *)
+  @objc
+  func collectReadings() -> Void {  
+    cancellable = WeatherSensorService.shared.reader.sensorReadingsPublisher
+      .sink{ value in
```
Comment (roberthartman) [inline ios/WeatherSensorModule.swift]: maybe I didn't notice this before, there should be a space between sink and { `sink {`

File: ios/WeatherSensorModule.swift [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+  @available(iOS 13.0, *)
+  @objc
+  func startSensorReadings() -> Void {
+    WeatherSensorService.shared.reader.startSensorReadings()
+  }
+
+  @available(iOS 13.0, *)
+  @objc
+  func stopSensorReadings() -> Void {
```
Comment (roberthartman) [inline ios/WeatherSensorModule.swift]: Even though it still compiles with the `-> Void`, you don't ever see that, in Swift a Void function just doesn't have a return type `func stopSensorReadings() {`. Please clean up the functions above too.

**Review by roberthartman — APPROVED** [review]
Approving, but please do the additional clean up that I commented on. Thanks!

File: android/src/main/java/com/reactnativeweathersensormodule/WeatherSensorModule.kt [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+    fun removeListeners(count: Int) {}
+
+    @ReactMethod
+    fun set(readingInterval: UInt) {
+      WeatherSensorService.reader.set(readingInterval)
+    }
+    
+    @ReactMethod 
+    fun startSensorReadings() {
+      WeatherSensorService.reader.startSensorReadings()
```
Comment (kevin-r-wilson) [inline android/src/main/java/com/reactnativeweathersensormodule/WeatherSensorModule.kt:50]: Are you not able to restart readings? Collect scope is cancelled on stop but it's not restarted.

**Review by kevin-r-wilson — APPROVED** [review]
Approved with one comment on restarting the scope to collect readings

File: android/src/main/java/com/reactnativeweathersensormodule/WeatherSensorModule.kt (hunk 21 above)
Comment (ClaireLundy) [reply android/src/main/java/com/reactnativeweathersensormodule/WeatherSensorModule.kt:50]: @kevin-r-wilson Calling startSensorReadings() starts sending them to the flow, you must then call collectReadings() to read from the flow. If you were to call stopSensorReadings, you would need to call startSensorReadings() again or calling collectReadings() would not return anything. I guess if you never called collectReadings() then the Collect Scope would be cancelled without ever being launched. Would that cause issues?
